### PR TITLE
fix udp service blackhole problem when number of backends changes from 0 to non-0

### LINKED ISSUE #48370: Services (kube-proxy): conntrack for UDP needs to be flushed when set of endpoints goes from empty to non-empty
# Summary

For the UDP protocol, we currently flush stale entries when the set of endpoints changes (update/remove). However, we do not flush the entries when the endpoint set changes from empty to non-empty.

When the endpoint set is empty, there is a conntrack entry that blackholes the traffic. If the conntrack entry is refreshed by new UDP activity, this entry will persist until the next conntrack flush. Note: the UDP socket has to be reused.

# How to repro:

Running in a pod to a service with ClusterIP 10.15.240.10 with one pod (10.12.0.11) as the backend :
```python
import socket, time

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
sock.bind(('0.0.0.0', 0))
print sock.getsockname()
while True:
  sock.sendto('abc', ('10.15.240.10', 53))
  time.sleep(1.0)
```

You will have conntrack entry:
```
udp      17 29 src=10.12.0.14 dst=10.15.240.10 sport=43145 dport=53 [UNREPLIED] src=10.12.0.11 dst=10.12.0.14 sport=53 dport=43145 mark=0 use=1
```
Change service to have zero endpoints by killing the pod, we get the following conntrack entry:
```
udp      17 29 src=10.12.0.14 dst=10.15.240.10 sport=43145 dport=53 [UNREPLIED] src=10.15.240.10 dst=10.12.0.14 sport=53 dport=43145 mark=0 use=1
```
After the pod restarts and is added back to the endpoints, the conntrack entry will never go away **IF** there is constant UDP activity on the existing socket.

# References
* Nginx DNS resolution breaks: https://github.com/kubernetes/kubernetes/issues/26309#issuecomment-258828324

### PR DESCRIPTION
**Pull request by freehan**
fixes: #48370

```release-note
Fix udp service blackhole problem when number of backends changes from 0 to non-0
```

### CODE REVIEW

**Comment by freehan** [comment]
adding cherrypick-candidate label because this can potential solve a bunch of "my kube-dns does not work" case.

**Comment by wojtek-t** [comment]
This looks sane to me, but I would prefer @bowei or @thockin to take a look.

**Comment by freehan** [comment]
I will fix the unit test and repush

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -755,16 +755,16 @@ func (proxier *Proxier) OnEndpointsSynced() {
 func updateEndpointsMap(
 	endpointsMap proxyEndpointsMap,
 	changes *endpointsChangeMap,
-	hostname string) (hcEndpoints map[types.NamespacedName]int, staleSet map[endpointServicePair]bool) {
-	staleSet = make(map[endpointServicePair]bool)
-
+	hostname string) (hcEndpoints map[types.NamespacedName]int, staleEndpoints map[endpointServicePair]bool, staleServiceNames map[proxy.ServicePortName]bool) {
```
Comment (bowei) [inline pkg/proxy/iptables/proxier.go]: Can we create struct for the return value so it's not so unwieldy 

```
type updateResult struct {
  healthcheckEndpoints map[types.NamespacedName]int
  staleEndpoints map[endpointServicePair]bool
  staleServices map[proxy.ServicePortName]bool
}
```

**Comment by thockin** [comment]
/approve

@bowei has LGTM

**Comment by shyamjvs** [comment]
/test pull-kubernetes-kubemark-e2e-gce
@wojtek-t @gmarek This should work now. IIUC the problem was this test was run before https://github.com/kubernetes/test-infra/pull/3309 was merged but after they migrated to passing --test-args to kubetest directly from e2e-runner.sh. So the "--test" arg was wrongly enabled.

**Comment by shyamjvs** [comment]
Seems like it is still running test against the old e2e-runner.sh:
```
W0707 01:53:29.702] +(/workspace/e2e-runner.sh:53): main(): e2e_go_args+=(--test)
```

As that line doesn't exist in [e2e-runner.sh](https://github.com/kubernetes/test-infra/blob/master/jenkins/e2e-image/e2e-runner.sh) anymore.
@fejta Seems like the kubekins image you pushed yesterday wasn't up to date. I'll push a new one and update the tag.

**Comment by shyamjvs** [comment]
/test pull-kubernetes-kubemark-e2e-gce
It should hopefully work now.

**Comment by shyamjvs** [comment]
/test pull-kubernetes-kubemark-e2e-gce
(Sorry for the spam, I'm on fixing it)

**Comment by freehan** [comment]
Ping

**Comment by wojtek-t** [comment]
@freehan - I'm fine with cherrypicking it to 1.7, but please add add a release note to this PR (describing the bug this is fixing).

**Comment by wojtek-t** [comment]
#48809 constains a cherrypick of it

**Pull request merged**